### ISSUE TITLE
Implement compliant-compliant hydroelastic contact

### ISSUE THREAD

**Issue by DamrongGuoy** [RobotLocomotion/drake#16040]
<h1>Target February 2022 Release</h1>
This is the implementation plan to change prototype #15545 into quality code in master.  It will be updated as we go. Each item should become a small PR. This is a team work with contributions from several members including @joemasterjohn and @rpoyner-tri.

#### ComputeContactSurfaceFromCompliantVolumes()
  - [X] Equilibrium plane between two linear functions from two tetrahedra. (@DamrongGuoy @rpoyner-tri )
    - #16052 
  - [x] Refactor mesh-plane intersection code for plane-tetrahedron intersection. (@joemasterjohn )
    - [x] #16458
  - [x] #16365 (@DamrongGuoy @joemasterjohn )
    - [x] #16374  
    - [X] Intersection of two linear-field tetrahdra with their equilibrium plane into a polygon.
    - [x] Heuristic to filter out special cases
         - Only allow contact polygons whose normal vectors are along pressure gradients. 
    - [x] All contact polygons between two tetrahedral meshes with pressure fields
    - [x] Contact surface between two compliant hydroelastic tetrahedral meshes

#### Use the new capability in SceneGraph
  - [x] Update hydroelastic_callback and proximity_engine_test (@DamrongGuoy @rpoyner-tri )
    - [x] #16366
        - Ready to merge, pending on `#16365`. 
        - [x] Check whether it breaks Anzu (Anzu-master-Drake-PR 147)
    - [x] Update doxygen of QueryObject::ComputeContactSurfaces() 
      - [x] #16471

#### Use the new capability in MultibodyPlant
  - [x] Update MultibodyPlant::CalcDiscreteContactPairs() (@DamrongGuoy @amcastro-tri @xuchenhan-tri )
    - [x]  #16367
      - [x] Confirm that it plays well with Anzu (Anzu-master-Drake-PR 149) :heavy_check_mark: 
  - [x] Confirm with @amcastro-tri that it works well with HydroelasticTractionCalculator and other parts.
    - @amcastro-tri gave the new formula for the effective hydroelastic pressure gradient between two objects. 

<h1> Half space </h1>
Move the request for compliant-halfspace contacting another compliant object to #16487.

**Comment by DamrongGuoy** [comment]
@hongkai-dai 

I have to postpone compliant-compliant hydroelastic contact to CY22Q1 since #15796 has a higher priority.  This issue tracks the progress of compliant-compliant hydroelastics.

**Comment by sherm1** [comment]
Congratulations @DamrongGuoy & team! 👏👏👏
